web_poet/page_inputs.py

```diff
@@ -1,17 +1,187 @@
+    and ``status``.
+
+    ``status`` should represent the int status code of the HTTP response.
+
+    ``headers`` should contain the HTTP response headers.
+
+    ``encoding`` encoding of the response. If None (default), encoding
+    is auto-detected from headers and body content.
     """
-    url: str
-    html: str
+
+    url: str = attrs.field()
+    body: HttpResponseBody = attrs.field(converter=HttpResponseBody)
+    status: Optional[int] = attrs.field(default=None)
+    headers: HttpResponseHeaders = attrs.field(factory=HttpResponseHeaders,
+                                               converter=HttpResponseHeaders)
+    _encoding: Optional[str] = attrs.field(default=None)
+
+    _DEFAULT_ENCODING = 'ascii'
+    _cached_text: Optional[str] = None
+
+    @property
+    def text(self) -> str:
+        """
+        Content of the HTTP body, converted to unicode
+        using the detected encoding of the response, according
+        to the web browser rules (respecting Content-Type header, etc.)
+        """
+        # Access self.encoding before self._cached_text, because
+        # there is a chance self._cached_text would be already populated
+        # while detecting the encoding
+        encoding = self.encoding
+        if self._cached_text is None:
+            fake_content_type_header = f'charset={encoding}'
+            encoding, text = html_to_unicode(fake_content_type_header, self.body)
+            self._cached_text = text
+        return self._cached_text
+
+    @property
+    def encoding(self):
+        """ Encoding of the response """
+        return (
+            self._encoding
+            or self._headers_declared_encoding()
+            or self._body_declared_encoding()
+            or self._body_inferred_encoding()
+        )
+
+    # XXX: see https://github.com/python/mypy/issues/1362
+    @property   # type: ignore
+    @memoizemethod_noargs
+    def selector(self) -> parsel.Selector:
+        # XXX: should we pass base_url=self.url, as Scrapy does?
+        return parsel.Selector(text=self.text)
+
+    def xpath(self, query, **kwargs):
+        return self.selector.xpath(query, **kwargs)
+
+    def css(self, query):
+        return self.selector.css(query)
+
+    @memoizemethod_noargs
+    def json(self):
+        return self.body.json()
+
+    @memoizemethod_noargs
+    def _headers_declared_encoding(self):
+        return self.headers.declared_encoding()
+
+    @memoizemethod_noargs
+    def _body_declared_encoding(self):
+        return self.body.declared_encoding()
+
+    @memoizemethod_noargs
+    def _body_inferred_encoding(self):
+        content_type = self.headers.get('Content-Type', '')
+        body_encoding, text = html_to_unicode(
+            content_type,
+            self.body,
+            auto_detect_fun=self._auto_detect_fun,
+            default_encoding=self._DEFAULT_ENCODING
+        )
+        self._cached_text = text
+        return body_encoding
+
+    def _auto_detect_fun(self, text):
```

Comment:
We could rename `text` as `body` and annotate it as `body: bytes` to be more clear.

Reply:
nice catch!